mmcv/cnn/bricks/transformer.py

```diff
@@ -36,35 +90,72 @@ def build_transformer_layer_sequence(cfg, default_args=None):
 
 
+def bnc_to_nbc(forward):
+    """This function can adjust the shape of dataflow('key', 'query', 'value')
+    from batch_first (batch, num_query, embed_dims) to num_query_first
+    (num_query ,batch, embed_dims)"""
+
+    def forward_wrapper(**kwargs):
+        convert_keys = ('key', 'query', 'value')
+        for key in kwargs.keys():
+            if key in convert_keys:
+                kwargs[key] = kwargs[key].transpose(0, 1)
+        attn_output, attn_output_weights = forward(**kwargs)
+        return attn_output.transpose(0, 1), attn_output_weights
+
+    return forward_wrapper
+
+
 @ATTENTION.register_module()
 class MultiheadAttention(BaseModule):
-    """A warpper for torch.nn.MultiheadAttention.
+    """A wrapper for Attention.
 
     This module implements MultiheadAttention with residual connection,
     and positional encoding used in DETR is also passed as input.
 
     Args:
         embed_dims (int): The embedding dimension.
-        num_heads (int): Parallel attention heads. Same as
-            `nn.MultiheadAttention`.
-        dropout (float):w A Dropout layer on attn_output_weights. Default: 0..
+        num_heads (int): Parallel attention heads.
+        attn_drop (float): A Dropout layer on attn_output_weights. Default 0.0.
+        dropout_layer (obj:`ConfigDict`): The dropout_layer used
+            when adding the shortcut.
         init_cfg (obj:`mmcv.ConfigDict`): The Config for initialization.
             Default: None.
+        batch_first (bool): Key, Query and Value are shape of
+            (batch, n, embed_dim)
+            or (n, batch, embed_dim). Default to False.
     """
 
     def __init__(self,
                  embed_dims,
                  num_heads,
-                 dropout=0.,
+                 attn_drop=0.,
+                 proj_drop=0.,
+                 dropout_layer=dict(type='DropOut', drop_prob=0.),
                  init_cfg=None,
+                 batch_first=False,
                  **kwargs):
         super(MultiheadAttention, self).__init__(init_cfg)
+        if 'dropout' in kwargs:
+            warnings.warn('The arguments `dropout` in MultiheadAttention '
+                          'has been deprecated, now you can separately '
+                          'set `attn_drop`(float), proj_drop(float), '
+                          'and `dropout_layer`(dict) ')
+            attn_drop = kwargs['dropout']
+            dropout_layer['drop_prob'] = kwargs.pop('dropout')
+
         self.embed_dims = embed_dims
         self.num_heads = num_heads
-        self.dropout = dropout
-        self.attn = nn.MultiheadAttention(embed_dims, num_heads, dropout,
+        self.batch_first = batch_first
+
+        self.attn = nn.MultiheadAttention(embed_dims, num_heads, attn_drop,
                                           **kwargs)
-        self.dropout = nn.Dropout(dropout)
+        if self.batch_first:
+            self.attn.forward = bnc_to_nbc(self.attn.forward)
```

Comment:
Such implementation is not elegant. I recommend using the [forward_pre hook](https://pytorch.org/docs/stable/generated/torch.nn.Module.html?highlight=forward_hook#torch.nn.Module.register_forward_pre_hook) here.

Reply:
I need transpose the dimension to (n,b,c) after forward

Reply:
Ok